lib/utils/ignore.ts

```diff
@@ -294,45 +339,65 @@ export async function filterFilesWithDockerignore(
 	const dockerignoreServiceDirs: string[] = multiDockerignore
 		? Object.keys(ignoreByDir).filter((dir) => dir && dir !== root)
 		: [];
+	const files = await listFiles(projectDir);
+
 	const dockerignoreFiles: FileStats[] = [];
 	const filteredFileList: FileStats[] = [];
-	const klaw = await import('klaw');
-
-	await new Promise((resolve, reject) => {
-		// Looking at klaw's source code, `preserveSymlinks` appears to only
-		// afect the `stats` argument to the `data` event handler
-		klaw(projectDir, { preserveSymlinks: false })
-			.on('error', reject)
-			.on('end', resolve)
-			.on('data', (item: { path: string; stats: Stats }) => {
-				const { path: filePath, stats } = item;
-				// With `preserveSymlinks: false`, filePath cannot be a symlink.
-				// filePath may be a directory or a regular or special file
-				if (!stats.isFile()) {
-					return;
-				}
+
+	// Because `files` can be of arbitrary length, process them in smaller batches
+	// to avoid out of memory errors.
+	// This approach is significantly faster than using Bluebird.map with a
+	// concurrency setting
+	const chunks = _.chunk(files, 750);
```

Comment:
Nice 👍  In your testing where you found that 750 was a sweet number, how many (logical/virtual) CPUs did the machine have? I am thinking of making this number proportional to the CPU count, something like:
```suggestion
	const chunks = _.chunk(files, Math.round(750 * (os.cpus().length || 1) / LUCIANS_CPU_COUNT));
```
Replace `LUCIANS_CPU_COUNT` with the result of `os.cpus().length` on the workstation where you ran the tests.

If the count on your machine was 16, then on a weaker device with cpu count 1 the chunk number would become 47.

I understand that Node.js is single threaded so CPU count would not be relevant in "user-space code", but I believe that the promises would be blocked on syscalls (`fs.stat`) that are executed by the kernel, and I assume that the kernel would take advantage of all CPU cores.

A similar calculation could be done in `listFiles`.

Reply:
@pdcastro My machine has 24 cores, but as you mention I'm not sure if the count is super relevant. Do you have any idea how to verify that `fs.stat` is affected by the number of CPU cores?

Reply:
I also wonder how necessary this change is, considering the main concern is out of memory errors. Perhaps we could merge the change and see if CPU consumption becomes an issue for people?